### PR TITLE
Add actions assembly to encode an action (i.e. approval) and its role, party, and approval date.

### PR DESCRIPTION
**Pull request by xee5ch**
# Committer Notes

A follow-up to usnistgov/OSCAL#1033, a recommended implementation for an `actions` assembly to allow developers using OSCAL to encode approval data for given staff certain `role`s for a `responsible-party` at a certain date and time.

This PR supercedes the obsolete PR usnistgov/OSCAL#1038.

### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/OSCAL/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/OSCAL/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you included examples of how to use your new feature(s)? (deferred to https://github.com/usnistgov/oscal-content/issues/130)
- [x] Have you updated all [OSCAL website](https://pages.nist.gov/OSCAL) and readme documentation affected by the changes you made? Changes to the OSCAL website can be made in the docs/content directory of your branch.

### CODE REVIEW

**Comment by wendellpiez** [comment]
Suggest:

* Add (required) `type` flag of type `token`. An example can show this with value "approval" (our primary use case).
* All other flags should be optional not required, should they not?

**Comment by xee5ch** [comment]
OK, feedback rolled in, and ready for final review. I hope to discuss tomorrow during model meeting if possible.

cc @david-waltermire-nist @wendellpiez

**Comment by david-waltermire-nist** [comment]
I believe we should adjust the action type to be namespaced similar to properties. This will allow different organizations to define new action types without coordinating with OSCAL. OSCAL can define a core set of actions in the OSCAL namespace.

**Comment by xee5ch** [comment]
Per feedback from Dave, I need to improve the documentation for the new assembly, fields, and flags before finalizing and removing the draft tag.

**Comment by xee5ch** [comment]
After a second pass, it seems adding other documentation strings will be too wordy and add unnecessary content. I think this is ready for the NIST OSCAL Team to review. :-)

File: src/metaschema/oscal_metadata_metaschema.xml [hunk 1]
```diff
@@ -751,6 +754,55 @@
...
+        </define-flag>
+        <define-flag name="date" as-type="dateTime-with-timezone">
+            <description>The date and time when the action occurred.</description>
+        </define-flag>
+        <define-flag name="type" as-type="token" required="yes">
+            <description>The type of action documented by the assembly, such as an approval.</description>
+        </define-flag>
+        <define-flag name="system" as-type="uri" required="yes">
+            <formal-name>Action Type System</formal-name>
+            <description>Specifies the action type system used.</description>
```
Comment (david-waltermire-nist) [inline src/metaschema/oscal_metadata_metaschema.xml:780]: We should be more specific about how namespace and type play together to define the actions "type" in an organizationally extensible way.

Comment (david-waltermire-nist) [reply src/metaschema/oscal_metadata_metaschema.xml:780]: Maybe base the text on similar language from prop or information-type.

Comment (xee5ch) [reply src/metaschema/oscal_metadata_metaschema.xml:783]: I updated it based on your feedback, @david-waltermire-nist, thanks.

**Comment by xee5ch** [comment]
OK, I think this is ready. I addressed feedback above in https://github.com/usnistgov/OSCAL/commit/fe39f58a92b5a8160b2b68f5b05e701ff1ef7dbc but also added some missing `formal-name`s and other things I should probably add. Let me know if you want further changes or that goes beyond the scope of what you think was/is prudent.

**Comment by xee5ch** [comment]
And woops, last amend was because of a typo in universally in one doc string, but now we should be good.

**Comment by wendellpiez** [comment]
If this is linked to `@ns` would it be better to call it `action/@name` not `action/@type`?

**Comment by xee5ch** [comment]
> If this is linked to `@ns` would it be better to call it `action/@name` not `action/@type`?

When we met to discuss the PR, Dave suggested that `@ns` was only appropriate with `@name` and maybe we should stick with `@type`. I guess he and others will weigh in here, but the only PR feedback was to touch up docs in the model and to move forward with a `@system="URI"` approach instead. See https://github.com/usnistgov/OSCAL/pull/1052/files#r943772853.

File: src/metaschema/oscal_metadata_metaschema.xml [hunk 2]
```diff
@@ -751,6 +754,62 @@
...
+            <description>The universally unique identifier of the party responsible for the action.</description>
+        </define-flag>
+        <define-flag name="date" as-type="dateTime-with-timezone">
+            <formal-name>Action Occurrence Date</formal-name>
+            <description>The date and time when the action occurred.</description>
+        </define-flag>
+        <define-flag name="type" as-type="token" required="yes">
+            <formal-name>Action Type</formal-name>
+            <description>The type of action documented by the assembly, such as an approval.</description>
+        </define-flag>
```
Comment (david-waltermire-nist) [inline src/metaschema/oscal_metadata_metaschema.xml:780]: Are there any initial built-in OSCAL types in the OSCAL system namespace? If so, we should add constraints for these.

File: src/metaschema/oscal_metadata_metaschema.xml [hunk 3]
```diff
@@ -751,6 +754,62 @@
...
+            <description>A unique identifier that can be used to reference this defined action elsewhere in an OSCAL document. A UUID should be consistently used for a given location across revisions of the document.</description>
+        </define-flag>
+        <define-flag name="role-id" as-type="token" required="yes">
+            <formal-name>Action Responsible Role Identifier</formal-name>
+            <description>Identifier of the approving role in the responsible party.</description>
+        </define-flag>
+        <define-flag name="party-uuid" as-type="token" required="yes">
+            <formal-name>Action Party Universally Unique Identifier</formal-name>
+            <description>The universally unique identifier of the party responsible for the action.</description>
+        </define-flag>
```
Comment (david-waltermire-nist) [inline src/metaschema/oscal_metadata_metaschema.xml:772]: For consistency, it might be better to use `responsible-party` here instead of these. Thoughts?

Comment (xee5ch) [reply src/metaschema/oscal_metadata_metaschema.xml:772]: I can agree to that.

File: src/metaschema/oscal_metadata_metaschema.xml (hunk 2 above)
Comment (xee5ch) [reply src/metaschema/oscal_metadata_metaschema.xml:780]: I am fine adding constraints. Is the only action for now going to be `approval`?

**Comment by xee5ch** [comment]
OK, I touched up with the feedback requests, @david-waltermire-nist and company. I am iffy on the `constraint`s but and XPathy syntax `target="system"` versus `target="./system"` et cetera. If we are good with it as-is, I will merge the commits as needed once your formal review is complete. :-)

**Comment by wendellpiez** [comment]
In XPath and also Metapath, `./system` is an abbreviated form of `self::node()/child::system` and returns the same nodes as `system` (expanded as `child::system`). So for purposes of addressing nodes they are functionally equivalent.

This doesn't mean that all XPath processors always treat them the same way, e.g. a delinter might even rewrite one to the other, etc.

**Comment by xee5ch** [comment]
Also, in terms of acceptance criteria, we still ask for examples. It seems in #1263 NIST has begun removing `example`s from the Metaschema and I am not sure of: do we need examples for this PR? Do we need a follow-up issue for examples? Either way, where should new examples go? Let me know. Thank you!

**Review by david-waltermire-nist — APPROVED** [review]
This looks good. Thanks!

**Comment by david-waltermire-nist** [comment]
> Also, in terms of acceptance criteria, we still ask for examples. It seems in #1263 NIST has begun removing `example`s from the Metaschema and I am not sure of: do we need examples for this PR? Do we need a follow-up issue for examples? Either way, where should new examples go? Let me know. Thank you!

It would be good to add an issue to create an example in the oscal-content repo maybe?

**Comment by aj-stein-nist** [comment]
@david-waltermire-nist, I pointed this PR to a feature branch as instructed during the weekly sync for when we are ready to merge after soliciting model feedback during an upcoming Model Review meeting next week.

**Review by david-waltermire-nist — APPROVED** [review]
LGTM. Thanks!
